### DISCUSSION TITLE
deep filters in populateWhere, results in unpopulated entities from other relation side

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#5302] [Q&A]
I have the following entities: `EntityA`, `EntityB`, `EntityC`, `EntityD`:

`EntityA` and `EntityB` have a 1:many relationship
`EntityA` and `EntityC` have a 1:many relationship
`EntityA` and `EntityD` have a many:1 relationship

I need to fetch `EntityB`s and populate `EntityA` with its corrosponding `EntityC`s, and `EntityD`s, but only populate `EntityC`s that match a certain condition. I want `EntityD` to be populated regardless.

Here's the query:

```ts
const result = await em.find(EntityB, {}, {
    populate: ["entityA", "entityA.entityD", "entityA.entityCs"],
    populateWhere: {entityA: {entityCs: {booleanProperty: true}}},
    logging: {enabled: true}
});
```
This works when there are `EntityC`s that match the condition, but when there aren't, `EntityD`s are not populated. Is this expected behavior or a bug? If it's not a bug what should be the query instead?

Here's the full repro:

```ts
import {Collection, Entity, LoadStrategy, ManyToOne, MikroORM, OneToMany, PrimaryKey, Property, Ref, types, wrap} from "@mikro-orm/sqlite";
import crypto from "node:crypto";

@Entity()
export class EntityA {
    @PrimaryKey({type: types.integer})
    id!: number;

    @OneToMany({entity: () => EntityB, mappedBy: (entityB: EntityB) => entityB.entityA})
    entityBs: Collection<EntityB> = new Collection<EntityB>(this);

    @OneToMany({entity: () => EntityC, mappedBy: (entityC: EntityC) => entityC.entityA})
    entityCs: Collection<EntityC> = new Collection<EntityC>(this);

    @ManyToOne({entity: () => EntityD, inversedBy: (entityD) => entityD.entityAs, deleteRule: "cascade", updateRule: "cascade"})
    entityD!: Ref<EntityD>;

}

@Entity()
export class EntityB {
    @PrimaryKey({type: types.integer})
    id!: number;

    @ManyToOne({entity: () => EntityA, inversedBy: (entityA) => entityA.entityBs, deleteRule: "cascade", updateRule: "cascade"})
    entityA!: EntityA;
}

@Entity()
export class EntityC {
    @PrimaryKey({type: types.integer})
    id!: number;

    @Property({type: types.boolean})
    booleanProperty!: boolean;

    @ManyToOne({entity: () => EntityA, inversedBy: (entityA) => entityA.entityCs, deleteRule: "cascade", updateRule: "cascade"})
    entityA!: EntityA;
}

@Entity()
export class EntityD {
    @PrimaryKey({type: types.integer})
    id!: number;

    @OneToMany({entity: () => EntityA, mappedBy: (entityA: EntityA) => entityA.entityD})
    entityAs: Collection<EntityA> = new Collection<EntityA>(this);
}


let orm: MikroORM;

beforeAll(async () => {
    orm = await MikroORM.init({
        dbName: ":memory:",
        entities: [EntityA, EntityB, EntityC, EntityD],
        loadStrategy: LoadStrategy.SELECT_IN,
        // debug: ["query", "query-params"],
        allowGlobalContext: true, // only for testing
    });
    await orm.schema.refreshDatabase();
});
beforeEach(async () => {
    await orm.schema.clearDatabase();
});
afterAll(async () => {
    await orm.close(true);
});

test("If there are entityCs that match the condition, entityD is loaded ", async () => {
    let em = orm.em.fork();
    const entityD = em.create(EntityD, {});
    const eA1 = em.create(EntityA, {entityD: entityD});
    const eB1 = em.create(EntityB, {entityA: eA1});
    const eCs = [...Array(5).keys()].map(i => em.create(EntityC, {entityA: eA1, booleanProperty: i % 2 == 0}));

    await em.flush();
    em.clear();
    em = orm.em.fork();
    const result = await em.find(EntityB, {}, {
        populate: ["entityA", "entityA.entityD", "entityA.entityCs"],
        populateWhere: {entityA: {entityCs: {booleanProperty: true}}},
        logging: {enabled: true}
    });
    expect(result[0].entityA.entityD).toBeDefined();
});

test("If there are no entityCs that match the condition, entityD is undefined", async () => {
    let em = orm.em.fork();
    const entityD = em.create(EntityD, {});
    const eA1 = em.create(EntityA, {entityD: entityD});
    const eB1 = em.create(EntityB, {entityA: eA1});
    const eCs = [...Array(5).keys()].map(_ => em.create(EntityC, {entityA: eA1, booleanProperty: false}));

    await em.flush();
    em.clear();
    em = orm.em.fork();
    const result = await em.find(EntityB, {}, {
        populate: ["entityA", "entityA.entityD", "entityA.entityCs"],
        populateWhere: {entityA: {entityCs: {booleanProperty: true}}},
        logging: {enabled: true}
    });
    expect(result[0].entityA.entityD).toBeDefined();
});


```

**Comment by B4nan** [comment]
This is a bit tricky use case. FYI it works fine with joined strategy (this is actually the only difference I am aware of right now). The problem with the select-in strategy is that your query results in two queries to load `entityA.entityD` and `entityA.entityCs`, both respecting the `populateWhere: { entityA: { entityCs: { booleanProperty: true } } }` as it says how to limit the `entityA` relation.

I don't think there is a way around this other than rewriting how you query the database - split the queries so you can use the `populateWhere` only for the second relation (which I believe is what you want in the first place):

```ts
const result = await em.find(EntityB, {}, {
  populate: ['entityA', 'entityA.entityD'],
  logging: { enabled: true },
});
// if you want to keep things type-safe for the loaded relations, use the return value of `em.populate`
await em.populate(result, ['entityA.entityCs'], {
  where: { entityA: { entityCs: { booleanProperty: true } } },
  logging: { enabled: true },
});
```

(or use the joined strategy)

edit: looking closer now, I think it actually limits the `entityA` relation already, it's not even trying to query its children as your `populateWhere` condition results in no `entityA` being loaded. if you `console.log(result[0].entityA);`, you can see its unloaded reference (because the populate query didnt return anything for it)

**Reply by squivix** [reply]
I thought [`populateWhere` had no effect on the Joined strategy](https://github.com/mikro-orm/mikro-orm/discussions/2988#discussioncomment-2510331)?  Or has that changed?

I tried to `console.log(wrap(result[0].entityA).isInitialized())` and it's false so I guess you're right.

I think maybe `populateWhere` should give more control over where the where filter should be applied. Maybe a feature request

**Reply by B4nan** [reply]
> I thought https://github.com/mikro-orm/mikro-orm/discussions/2988#discussioncomment-2510331? Or has that changed?

Yes, that changed with 5.8 (https://mikro-orm.io/blog/mikro-orm-5-8-released#filters-with-joined-loading-strategy). And since v6, joined strategy should (mostly) behave the same as select-in, the only difference is basically what we are discussing now - loading multiple levels of relations with conditions on the children. We could probably unify this too with one trick, but the unification would mean joined strategy will behave like the select-in now, not the other way around.

> I tried to console.log(wrap(result[0].entityA).isInitialized()) and it's false so I guess you're right.

You can tell just by logging the entity, when you see it logged in parens, it means it's not loaded:

```ts
(EntityA) { id: 1 }
```

> I think maybe populateWhere should give more control over where the where filter should be applied. Maybe a feature request

How would you expect the API to look?

**Reply by squivix** [reply]
> How would you expect the API to look?

I don't know. Maybe an object with key being one of the `populate` elements and value having a where property. For example:

```ts
const result = await em.find(EntityB, {}, {
    populate: ["entityA", "entityA.entityD", "entityA.entityCs"],
    populateWhere: {
        "entityA.entityCs": {where: {booleanProperty: true}},
        logging: {enabled: true}
    }
});
```

It's just a suggestion. For now, the seperate `em.populate()` call works nicely for me.
